functions/src/winloss.ts

```diff
@@ -0,0 +1,122 @@
+import { firestore } from "firebase-admin";
+import * as functions from "firebase-functions";
+import { db } from "./app";
+import { getPrivatePlayerCollection, lobbyCollection } from "./firestore-collections";
+import { isLobbyRequest } from "./firebase-functions-types";
+import { Lobby, PrivatePlayer } from "./firestore-types/lobby";
+
+// this function will likely be obsolete once timers are introduced
+export const lobbyReturn = functions.https.onCall(async (data: unknown, context): Promise<void> => {
+  if (context.auth === undefined) {
+    throw new functions.https.HttpsError("permission-denied", "Not Signed In");
+  }
+
+  // validate code
+  if (!isLobbyRequest(data)) {
+    throw new functions.https.HttpsError("invalid-argument", "Invalid lobby code!");
+  }
+
+  // get lobby doc
+  const lobby = await lobbyCollection.doc(data.code).get();
+  if (lobby.exists === false) {
+    throw new functions.https.HttpsError("not-found", "Lobby doesn't exist!");
+  }
+
+  const { uids, state } = lobby.data() as Lobby;
+  // check if host
+  if (context.auth.uid !== uids[0]) {
+    throw new functions.https.HttpsError("permission-denied", "Not the host of the game!");
+  }
+
+  // check game state
+  if (state !== "END") {
+    throw new functions.https.HttpsError("invalid-argument", "Wrong game state!");
+  }
+
+  await resetLobby(lobby.ref);
+  await deletePrivatePlayers(lobby.ref, getPrivatePlayerCollection(lobby.ref));
+});
+
+export function findWinner(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobbyData === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+    const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+    const { players, uids } = lobbyData as Lobby;
+
+    for (let i = 0; i < players.length; i++) {
+      const privatePlayerDocRef = privatePlayerCollection.doc(uids[i]);
+      const privatePlayerDoc = await transaction.get(privatePlayerDocRef);
```

Comment:
an `await` inside a loop can be really slow maybe consider using [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) and [Array.map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)